# Networks: structure, evolution & processes
**Internet Analytics - Lab 2**

---

**Group:** *E*

**Names:**

* *Enrico Benedettini*
* *Mak Fazlic*
* *Ali Essonni*
* *Youssef Amine*

---

#### Instructions

*This is a template for part 2 of the lab. Clearly write your answers, comments and interpretations in Markodown cells. Don't forget that you can add $\LaTeX$ equations in these cells. Feel free to add or remove any cell.*

*Please properly comment your code. Code readability will be considered for grading. To avoid long cells of codes in the notebook, you can also embed long python functions and classes in a separate module. Don’t forget to hand in your module if that is the case. In multiple exercises, you are required to come up with your own method to solve various problems. Be creative and clearly motivate and explain your methods. Creativity and clarity will be considered for grading.*

---

## 2.2 Network sampling

#### Exercise 2.7: Random walk on the Facebook network

In [1]:
import math
import requests
import random
import numpy as np

In [2]:
def getData(node):
    URL_TEMPLATE = 'http://iccluster050.iccluster.epfl.ch:5050/v1.0/facebook?user={user_id}'
    url = URL_TEMPLATE.format(user_id=node)
    response = requests.get(url)
    return response.json()

def walker(starting_node, number_nodes):
    i = 0
    node = starting_node
    ages = []
    while i <number_nodes:
        data = getData(node)
        ages.append(data["age"])
        neighboors = data["friends"]
        id = random.randrange(0, len(neighboors))
        node = neighboors[id]
        i = i + 1
    return np.sum(ages) / number_nodes

In [ ]:
user_id = "a5771bce93e200c36f7cd9dfd0e5deaa"
users = 1000
avg = walker(user_id,users)

In [101]:
print(f"The average age of facebook users using the user {user_id} as a starting node and using {users} is {avg} years")

The average age of facebook users using the user a5771bce93e200c36f7cd9dfd0e5deaa as a starting node and using 1000 is 25.223 years


#### Exercise 2.8

1. We got 25.2 years as average years which  is not close to the true average which is 45 by 19.8 years
2. The discrepency can be explained by the fact if we have user a and he has user b c d in his friends list and we take randomly user b then, user b has user a in his friend list and can be taken by the random generator and creates mistakes in the average that we are computing

In [3]:
#3.
def walker_upgraded(starting_node, number_nodes):
    ages = []
    degrees = []
    n = number_nodes
    node = starting_node
    while n > 0:
        data = getData(node)
        neighboors = data["friends"]
        age = data["age"]
        deg = len(neighboors)
        id = random.randrange(0, len(neighboors))
        node = neighboors[id]
        n = n - 1
        ages.append(age)
        degrees.append(deg)

    ages = np.array(ages)
    degrees = np.array(degrees)
    return np.sum(ages / degrees) / np.sum(1 / degrees)

In [6]:
user_id = "a5771bce93e200c36f7cd9dfd0e5deaa"
users = 10000
avg = walker_upgraded(user_id,users)

In [7]:
print(avg)

44.87313458514549
